File: scripts/validate-project-memory.py
#!/usr/bin/env python3
"""Validate a project-memory.json document against the repo schema.

Usage:
  python3 scripts/validate-project-memory.py [path]
  python3 scripts/validate-project-memory.py --self-test

The schema (see docs/templates/project-memory.json):

  - version (int, currently 1)
  - task (string, may be empty)
  - techStack (object: languages[], frameworks[], tooling[])
  - build (object: install, test, lint, typecheck — all strings)
  - conventions (object: indent, filenameCase, moduleLayout — all strings)
  - structure (object: src, tests, docs — all strings)
  - userOwned (object: customNotes[], directives[] — both string arrays)
  - hotPaths (string array)

``userOwned.customNotes`` and ``userOwned.directives`` are user-protected:
the validator does not enforce content, only the array-of-strings shape,
so automated agents can safely append to them without overwriting prior
entries.

``--self-test`` mode is fully isolated under ``tempfile.TemporaryDirectory``
and seeds one clean fixture plus four negative fixtures.
"""
from __future__ import annotations

import json
import sys
import tempfile
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
TEMPLATE_PATH = ROOT / "docs" / "templates" / "project-memory.json"

ALLOWED_TOP_KEYS = {
    "$comment",
    "version",
    "task",
    "techStack",
    "build",
    "conventions",
    "structure",
    "userOwned",
    "hotPaths",
}
ALLOWED_TECHSTACK_KEYS = {"languages", "frameworks", "tooling"}
ALLOWED_BUILD_KEYS = {"install", "test", "lint", "typecheck"}
ALLOWED_CONVENTIONS_KEYS = {"indent", "filenameCase", "moduleLayout"}
ALLOWED_STRUCTURE_KEYS = {"src", "tests", "docs"}
ALLOWED_USEROWNED_KEYS = {"$comment", "customNotes", "directives"}


def _fail(message: str) -> None:
    print(f"FAIL: {message}", file=sys.stderr)
    raise SystemExit(1)


def _ok(message: str) -> None:
    print(f"ok: {message}")


def _require_keys(obj: dict, allowed: set[str], label: str) -> None:
    extra = set(obj.keys()) - allowed
    if extra:
        _fail(f"{label} contains unsupported fields: {sorted(extra)}")


def _require_str_array(obj: object, label: str) -> None:
    if not isinstance(obj, list):
        _fail(f"{label} must be an array")
    for index, item in enumerate(obj):
        if not isinstance(item, str):
            _fail(f"{label}[{index}] must be a string")


def _require_str_dict(obj: object, allowed: set[str], label: str) -> None:
    if not isinstance(obj, dict):
        _fail(f"{label} must be an object")
    _require_keys(obj, allowed, label)
    for key, value in obj.items():
        if not isinstance(value, str):
            _fail(f"{label}.{key} must be a string")


def validate(path: Path) -> None:
    if not path.is_file():
        _fail(f"project memory file not found: {path}")
    try:
        text = path.read_text(encoding="utf-8")
    except OSError as exc:
        _fail(f"could not read {path}: {exc}")
    try:
        data = json.loads(text)
    except json.JSONDecodeError as exc:
        _fail(f"project memory is not valid JSON: {exc}")
    if not isinstance(data, dict):
        _fail("project memory root must be a JSON object")

    _require_keys(data, ALLOWED_TOP_KEYS, "project-memory")

    for required in ("version", "task", "techStack", "build", "conventions", "structure", "userOwned", "hotPaths"):
        if required not in data:
            _fail(f"project-memory missing required key: {required}")

    if not isinstance(data["version"], int) or data["version"] != 1:
        _fail("version must be the integer 1 (only schema version supported)")
    if not isinstance(data["task"], str):
        _fail("task must be a string (may be empty)")

    tech_stack = data["techStack"]
    if not isinstance(tech_stack, dict):
        _fail("techStack must be an object")
    _require_keys(tech_stack, ALLOWED_TECHSTACK_KEYS, "techStack")
    for key in ALLOWED_TECHSTACK_KEYS:
        if key not in tech_stack:
            _fail(f"techStack missing required key: {key}")
        _require_str_array(tech_stack[key], f"techStack.{key}")

    _require_str_dict(data["build"], ALLOWED_BUILD_KEYS, "build")
    for key in ALLOWED_BUILD_KEYS:
        if key not in data["build"]:
            _fail(f"build missing required key: {key}")

    _require_str_dict(data["conventions"], ALLOWED_CONVENTIONS_KEYS, "conventions")
    for key in ALLOWED_CONVENTIONS_KEYS:
        if key not in data["conventions"]:
            _fail(f"conventions missing required key: {key}")

    _require_str_dict(data["structure"], ALLOWED_STRUCTURE_KEYS, "structure")
    for key in ALLOWED_STRUCTURE_KEYS:
        if key not in data["structure"]:
            _fail(f"structure missing required key: {key}")

    user_owned = data["userOwned"]
    if not isinstance(user_owned, dict):
        _fail("userOwned must be an object")
    _require_keys(user_owned, ALLOWED_USEROWNED_KEYS, "userOwned")
    for key in ("customNotes", "directives"):
        if key not in user_owned:
            _fail(f"userOwned missing required key: {key}")
        _require_str_array(user_owned[key], f"userOwned.{key}")

    _require_str_array(data["hotPaths"], "hotPaths")

    _ok(f"project memory is valid: {path}")


# ---------------------------------------------------------------------------
# --self-test
# ---------------------------------------------------------------------------


_CLEAN_FIXTURE: dict = {
    "version": 1,
    "task": "",
    "techStack": {"languages": ["python"], "frameworks": [], "tooling": ["pytest"]},
    "build": {"install": "", "test": "pytest", "lint": "", "typecheck": ""},
    "conventions": {"indent": "4 spaces", "filenameCase": "snake_case", "moduleLayout": ""},
    "structure": {"src": "src/", "tests": "tests/", "docs": "docs/"},
    "userOwned": {"customNotes": ["note one"], "directives": ["never push to main"]},
    "hotPaths": ["scripts/validate-workflow-state.py"],
}


def _run_self_test() -> int:
    with tempfile.TemporaryDirectory() as td:
        sandbox = Path(td)

        clean = sandbox / "clean.json"
        clean.write_text(json.dumps(_CLEAN_FIXTURE, indent=2), encoding="utf-8")
        validate(clean)

        bad_version = sandbox / "bad-version.json"
        bad = dict(_CLEAN_FIXTURE)
        bad["version"] = 2
        bad_version.write_text(json.dumps(bad), encoding="utf-8")
        try:
            validate(bad_version)
        except SystemExit as exc:
            if exc.code != 1:
                print(f"FAIL: bad-version exit was {exc.code}", file=sys.stderr)
                return 1
            _ok("self-test detected bad-version")
        else:
            print("FAIL: bad-version not detected", file=sys.stderr)
            return 1

        bad_missing = sandbox / "missing-userowned.json"
        bad = dict(_CLEAN_FIXTURE)
        del bad["userOwned"]
        bad_missing.write_text(json.dumps(bad), encoding="utf-8")
        try:
            validate(bad_missing)
        except SystemExit as exc:
            if exc.code != 1:
                return 1
            _ok("self-test detected missing-userowned")
        else:
            print("FAIL: missing-userowned not detected", file=sys.stderr)
            return 1

        bad_type = sandbox / "bad-type.json"
        bad = json.loads(json.dumps(_CLEAN_FIXTURE))
        bad["techStack"]["languages"] = "python"  # type: ignore[index]
        bad_type.write_text(json.dumps(bad), encoding="utf-8")
        try:
            validate(bad_type)
        except SystemExit as exc:
            if exc.code != 1:
                return 1
            _ok("self-test detected bad-type")
        else:
            print("FAIL: bad-type not detected", file=sys.stderr)
            return 1

        bad_extra = sandbox / "bad-extra.json"
        bad = json.loads(json.dumps(_CLEAN_FIXTURE))
        bad["surprise"] = "field"
        bad_extra.write_text(json.dumps(bad), encoding="utf-8")
        try:
            validate(bad_extra)
        except SystemExit as exc:
            if exc.code != 1:
                return 1
            _ok("self-test detected bad-extra")
        else:
            print("FAIL: bad-extra not detected", file=sys.stderr)
            return 1

    if TEMPLATE_PATH.is_file():
        validate(TEMPLATE_PATH)
        _ok(f"shipped template passes: {TEMPLATE_PATH.relative_to(ROOT)}")
    else:
        print(f"FAIL: shipped template missing: {TEMPLATE_PATH}", file=sys.stderr)
        return 1

    print("VALIDATE_PROJECT_MEMORY_SELF_TEST_OK")
    return 0


def main(argv: list[str] | None = None) -> int:
    args = list(sys.argv[1:] if argv is None else argv)
    if "--self-test" in args:
        return _run_self_test()
    target = Path(args[0]) if args else TEMPLATE_PATH
    validate(target)
    print("VALIDATE_PROJECT_MEMORY_OK")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/validate-project-memory.ts
#!/usr/bin/env node
/**
 * Validate a project-memory.json document against the repo schema.
 *
 * Usage:
 *   node --experimental-strip-types scripts/validate-project-memory.ts [path]
 *   node --experimental-strip-types scripts/validate-project-memory.ts --self-test
 *
 * The schema (see docs/templates/project-memory.json):
 *
 *   - version (int, currently 1)
 *   - task (string, may be empty)
 *   - techStack (object: languages[], frameworks[], tooling[])
 *   - build (object: install, test, lint, typecheck — all strings)
 *   - conventions (object: indent, filenameCase, moduleLayout — all strings)
 *   - structure (object: src, tests, docs — all strings)
 *   - userOwned (object: customNotes[], directives[] — both string arrays)
 *   - hotPaths (string array)
 *
 * `userOwned.customNotes` and `userOwned.directives` are user-protected:
 * the validator does not enforce content, only the array-of-strings shape,
 * so automated agents can safely append to them without overwriting prior
 * entries.
 *
 * `--self-test` mode is fully isolated under `fs.mkdtempSync(os.tmpdir())`
 * and seeds one clean fixture plus four negative fixtures.
 */
import * as fs from 'node:fs';
import * as os from 'node:os';
import * as path from 'node:path';
import * as process from 'node:process';
import { fileURLToPath } from 'node:url';

const currentFile = fileURLToPath(import.meta.url);
const ROOT = path.resolve(path.dirname(currentFile), '..');
const TEMPLATE_PATH = path.join(ROOT, 'docs', 'templates', 'project-memory.json');

const ALLOWED_TOP_KEYS = new Set([
  '$comment',
  'version',
  'task',
  'techStack',
  'build',
  'conventions',
  'structure',
  'userOwned',
  'hotPaths',
]);
const ALLOWED_TECHSTACK_KEYS = new Set(['languages', 'frameworks', 'tooling']);
const ALLOWED_BUILD_KEYS = new Set(['install', 'test', 'lint', 'typecheck']);
const ALLOWED_CONVENTIONS_KEYS = new Set(['indent', 'filenameCase', 'moduleLayout']);
const ALLOWED_STRUCTURE_KEYS = new Set(['src', 'tests', 'docs']);
const ALLOWED_USEROWNED_KEYS = new Set(['$comment', 'customNotes', 'directives']);

class ValidationError extends Error {}

function fail(message: string): never {
  throw new ValidationError(message);
}

function ok(message: string): void {
  console.log(`ok: ${message}`);
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function requireKeys(obj: Record<string, unknown>, allowed: Set<string>, label: string): void {
  const extra = Object.keys(obj).filter(k => !allowed.has(k)).sort();
  if (extra.length > 0) {
    const formatted = '[' + extra.map(k => `'${k}'`).join(', ') + ']';
    fail(`${label} contains unsupported fields: ${formatted}`);
  }
}

function requireStrArray(obj: unknown, label: string): void {
  if (!Array.isArray(obj)) {
    fail(`${label} must be an array`);
  }
  for (let index = 0; index < obj.length; index++) {
    if (typeof obj[index] !== 'string') {
      fail(`${label}[${index}] must be a string`);
    }
  }
}

function requireStrDict(obj: unknown, allowed: Set<string>, label: string): void {
  if (!isPlainObject(obj)) {
    fail(`${label} must be an object`);
  }
  requireKeys(obj, allowed, label);
  for (const [key, value] of Object.entries(obj)) {
    if (typeof value !== 'string') {
      fail(`${label}.${key} must be a string`);
    }
  }
}

function validate(filePath: string): void {
  let stat: fs.Stats;
  try {
    stat = fs.statSync(filePath);
  } catch {
    fail(`project memory file not found: ${filePath}`);
  }
  if (!stat.isFile()) {
    fail(`project memory file not found: ${filePath}`);
  }

  let text: string;
  try {
    text = fs.readFileSync(filePath, 'utf-8');
  } catch (exc: any) {
    fail(`could not read ${filePath}: ${exc?.message || exc}`);
  }

  let data: unknown;
  try {
    data = JSON.parse(text);
  } catch (exc: any) {
    fail(`project memory is not valid JSON: ${exc?.message || exc}`);
  }

  if (!isPlainObject(data)) {
    fail('project memory root must be a JSON object');
  }

  requireKeys(data, ALLOWED_TOP_KEYS, 'project-memory');

  for (const required of ['version', 'task', 'techStack', 'build', 'conventions', 'structure', 'userOwned', 'hotPaths']) {
    if (!(required in data)) {
      fail(`project-memory missing required key: ${required}`);
    }
  }

  if (typeof data.version !== 'number' || !Number.isInteger(data.version) || data.version !== 1) {
    fail('version must be the integer 1 (only schema version supported)');
  }
  if (typeof data.task !== 'string') {
    fail('task must be a string (may be empty)');
  }

  const techStack = data.techStack;
  if (!isPlainObject(techStack)) {
    fail('techStack must be an object');
  }
  requireKeys(techStack, ALLOWED_TECHSTACK_KEYS, 'techStack');
  for (const key of ALLOWED_TECHSTACK_KEYS) {
    if (!(key in techStack)) {
      fail(`techStack missing required key: ${key}`);
    }
    requireStrArray(techStack[key], `techStack.${key}`);
  }

  requireStrDict(data.build, ALLOWED_BUILD_KEYS, 'build');
  for (const key of ALLOWED_BUILD_KEYS) {
    if (!(key in (data.build as Record<string, unknown>))) {
      fail(`build missing required key: ${key}`);
    }
  }

  requireStrDict(data.conventions, ALLOWED_CONVENTIONS_KEYS, 'conventions');
  for (const key of ALLOWED_CONVENTIONS_KEYS) {
    if (!(key in (data.conventions as Record<string, unknown>))) {
      fail(`conventions missing required key: ${key}`);
    }
  }

  requireStrDict(data.structure, ALLOWED_STRUCTURE_KEYS, 'structure');
  for (const key of ALLOWED_STRUCTURE_KEYS) {
    if (!(key in (data.structure as Record<string, unknown>))) {
      fail(`structure missing required key: ${key}`);
    }
  }

  const userOwned = data.userOwned;
  if (!isPlainObject(userOwned)) {
    fail('userOwned must be an object');
  }
  requireKeys(userOwned, ALLOWED_USEROWNED_KEYS, 'userOwned');
  for (const key of ['customNotes', 'directives']) {
    if (!(key in userOwned)) {
      fail(`userOwned missing required key: ${key}`);
    }
    requireStrArray(userOwned[key], `userOwned.${key}`);
  }

  requireStrArray(data.hotPaths, 'hotPaths');

  ok(`project memory is valid: ${filePath}`);
}

// ---------------------------------------------------------------------------
// --self-test
// ---------------------------------------------------------------------------

const CLEAN_FIXTURE = {
  version: 1,
  task: '',
  techStack: { languages: ['python'], frameworks: [], tooling: ['pytest'] },
  build: { install: '', test: 'pytest', lint: '', typecheck: '' },
  conventions: { indent: '4 spaces', filenameCase: 'snake_case', moduleLayout: '' },
  structure: { src: 'src/', tests: 'tests/', docs: 'docs/' },
  userOwned: { customNotes: ['note one'], directives: ['never push to main'] },
  hotPaths: ['scripts/validate-workflow-state.py'],
};

function tryValidateExpectingFailure(filePath: string, label: string): boolean {
  try {
    validate(filePath);
  } catch (exc) {
    if (exc instanceof ValidationError) {
      ok(`self-test detected ${label}`);
      return true;
    }
    throw exc;
  }
  console.error(`FAIL: ${label} not detected`);
  return false;
}

function runSelfTest(): number {
  const sandbox = fs.mkdtempSync(path.join(os.tmpdir(), 'validate-project-memory-'));
  try {
    const clean = path.join(sandbox, 'clean.json');
    fs.writeFileSync(clean, JSON.stringify(CLEAN_FIXTURE, null, 2), 'utf-8');
    validate(clean);

    const badVersion = path.join(sandbox, 'bad-version.json');
    const badVersionData = { ...CLEAN_FIXTURE, version: 2 };
    fs.writeFileSync(badVersion, JSON.stringify(badVersionData), 'utf-8');
    if (!tryValidateExpectingFailure(badVersion, 'bad-version')) return 1;

    const badMissing = path.join(sandbox, 'missing-userowned.json');
    const badMissingData: Record<string, unknown> = { ...CLEAN_FIXTURE };
    delete badMissingData.userOwned;
    fs.writeFileSync(badMissing, JSON.stringify(badMissingData), 'utf-8');
    if (!tryValidateExpectingFailure(badMissing, 'missing-userowned')) return 1;

    const badType = path.join(sandbox, 'bad-type.json');
    const badTypeData = JSON.parse(JSON.stringify(CLEAN_FIXTURE));
    badTypeData.techStack.languages = 'python';
    fs.writeFileSync(badType, JSON.stringify(badTypeData), 'utf-8');
    if (!tryValidateExpectingFailure(badType, 'bad-type')) return 1;

    const badExtra = path.join(sandbox, 'bad-extra.json');
    const badExtraData = JSON.parse(JSON.stringify(CLEAN_FIXTURE));
    badExtraData.surprise = 'field';
    fs.writeFileSync(badExtra, JSON.stringify(badExtraData), 'utf-8');
    if (!tryValidateExpectingFailure(badExtra, 'bad-extra')) return 1;
  } finally {
    fs.rmSync(sandbox, { recursive: true, force: true });
  }

  if (fs.existsSync(TEMPLATE_PATH) && fs.statSync(TEMPLATE_PATH).isFile()) {
    validate(TEMPLATE_PATH);
    ok(`shipped template passes: ${path.relative(ROOT, TEMPLATE_PATH)}`);
  } else {
    console.error(`FAIL: shipped template missing: ${TEMPLATE_PATH}`);
    return 1;
  }

  console.log('VALIDATE_PROJECT_MEMORY_SELF_TEST_OK');
  return 0;
}

function main(argv?: string[]): number {
  const args = argv ?? process.argv.slice(2);
  if (args.includes('--self-test')) {
    return runSelfTest();
  }
  const target = args.length > 0 ? args[0] : TEMPLATE_PATH;
  try {
    validate(target);
  } catch (exc) {
    if (exc instanceof ValidationError) {
      console.error(`FAIL: ${exc.message}`);
      return 1;
    }
    throw exc;
  }
  console.log('VALIDATE_PROJECT_MEMORY_OK');
  return 0;
}

process.exit(main());
